cornac/models/dream/dream.py

```diff
@@ -0,0 +1,375 @@
+import itertools
+import random
+
+import numpy as np
+import torch
+import torch.nn as nn
+
+from torch.autograd import Variable
+from torch.nn.init import constant_, xavier_normal_
+from torch.nn.utils.clip_grad import clip_grad_norm_
+from tqdm.auto import trange
+
+
+class Wloss(nn.modules.loss._Loss):
```

Comment:
What is W loss? Can it be more informative or is it an acronym by the authors?

Reply:
@tqtg I am not quite sure about the acronym for `Wloss`. The current implementation is based on this implemenation https://github.com/liming-7/A-Next-Basket-Recommendation-Reality-Check/tree/main/methods/dream

The authors of the above implementation also refer the running source code to this repo https://github.com/yihong-chen/DREAM

However, these two implementations are quite different in their naming convention. I haven't read through them thoroughly. Let's take some time to check the validity of both implementations, whether they reflect the [paper](https://cseweb.ucsd.edu/classes/fa17/cse291-b/reading/A%20Dynamic%20Recurrent%20Model%20for%20Next%20Basket%20Recommendation.pdf) idea.

Reply:
Glancing at the paper, I only see BPR loss is used without additional information. I think it's better to rely on the [original implementation](https://github.com/yihong-chen/DREAM/blob/master/src/train.py) which contains 2 variants (BPR and reordered BPR).